wallet/src/store.rs

```diff
@@ -0,0 +1,1196 @@
+                let mut index_lock_hashes = HashSet::new();
+                index_lock_hashes.insert(lock_hash.to_owned());
+                (from_block_number..=chain_state.tip_number()).for_each(|block_number| {
+                    let block = self
+                        .shared
+                        .block_hash(block_number)
+                        .and_then(|hash| self.shared.block(&hash))
+                        .expect("block exists");
+                    self.attach_block(batch, &index_lock_hashes, &block);
+                });
+            }
+            batch.insert_lock_hash_index_state(lock_hash, &index_state);
+        });
+        index_state
+    }
+
+    fn remove_lock_hash(&self, lock_hash: &H256) {
+        self.commit_batch(|batch| {
+            let iter = self
+                .db
+                .iter(COLUMN_LOCK_HASH_LIVE_CELL, lock_hash.as_bytes())
+                .expect("wallet db iter should be ok");
+
+            iter.take_while(|(key, _)| key.starts_with(lock_hash.as_bytes()))
+                .for_each(|(key, _)| {
+                    let lock_hash_index = LockHashIndex::from_slice(&key);
+                    batch.delete_lock_hash_live_cell(&lock_hash_index);
+                    batch.delete_cell_out_point_lock_hash(&lock_hash_index.cell_out_point);
+                });
+
+            let iter = self
+                .db
+                .iter(COLUMN_LOCK_HASH_TRANSACTION, lock_hash.as_bytes())
+                .expect("wallet db iter should be ok");
+
+            iter.take_while(|(key, _)| key.starts_with(lock_hash.as_bytes()))
+                .for_each(|(key, _)| {
+                    let lock_hash_index = LockHashIndex::from_slice(&key);
+                    batch.delete_lock_hash_transaction(&lock_hash_index);
+                });
+
+            batch.delete_lock_hash_index_state(&lock_hash);
+        });
+    }
+}
+
+impl<CS: ChainStore + 'static> DefaultWalletStore<CS> {
+    pub fn new(db_config: &DBConfig, shared: Shared<CS>) -> Self {
+        let db = RocksDB::open(db_config, COLUMNS);
+        DefaultWalletStore {
+            db: Arc::new(db),
+            shared,
+        }
+    }
+
+    pub fn start<S: ToString>(self, thread_name: Option<S>, notify: &NotifyController) {
+        let mut thread_builder = thread::Builder::new();
+        if let Some(name) = thread_name {
+            thread_builder = thread_builder.name(name.to_string());
+        }
+
+        let new_tip_receiver = notify.subscribe_new_tip(WALLET_STORE_SUBSCRIBER);
+        thread_builder
+            .spawn(move || loop {
+                select! {
+                    recv(new_tip_receiver) -> msg => match msg {
+                        Ok(tip_changes) => self.update(&tip_changes.detached_blocks, &tip_changes.attached_blocks),
+                        _ => {
+                            error!(target: "wallet", "new_tip_receiver closed");
+                            break;
```

Comment:
Correct me if I'm wrong: does this mean that at the first error, this thread would stop processing more tip updates? If so, I feel like this is too big an assumption.

Reply:
this error only occurs while channel closing, which means that main thread exits unexpectedly.